File: 2020/day3.py

input = []
for line in open('day3.txt', 'r'):
  input.append(line.strip())

def countTrees(input, slopes):
  w = len(input[0])
  treecount = 1

  for s in slopes:
    x = y = trees = 0
    for i in range(len(input)):
      if y == i:
        # print("i:",i,"x:",x,"y:",y,"w:",w,"l:",input[i],"=",input[i][x % w])
        trees += input[i][x % w] == '#'
        x += s[0]
        y += s[1]
    treecount *= trees

  return treecount

print("Part 1:", countTrees(input, [ [3,1] ]))
print("Part 2:", countTrees(input, [ [1,1], [3,1], [5,1], [7,1], [1,2] ]))

File: 2020/day3.rb

def getInput(t)
  return File.open("day3.#{t}").readlines()
end

def getSlopes(input, slopes)
  w = input[0].chomp.length
  treeCount = []
  slopes.each do |slope|
    x = y = trees = 0
    # puts "slope: #{slope}"
    input.each_with_index do |line, i|
      if y <= input.length && y == i
        # puts "i:#{i}, x:#{x}, y:#{y}, pos:#{x % w}, l: #{line.chomp} = #{line[x % w]}"
        trees += 1 if line[x % w] == '#'
        x += slope[0]
        y += slope[1]
      end
    end
    treeCount << trees
  end
  pp treeCount
  return treeCount.reduce(:*)
end

puts "Part 1: #{getSlopes(getInput('txt'), [ [1, 1], [3, 1], [5, 1], [7, 1], [1, 2] ])}"
